proxy/envoy/discovery.go

```diff
@@ -125,101 +173,178 @@ func (ds *DiscoveryService) Run() {
-		errorResponse(fmt.Sprintf("Unexpected %s %q", ServiceCluster, serviceCluster), response)
-		return
-	}
+	ds.cacheMu.Lock()
+	defer ds.cacheMu.Unlock()
 
-	// service-node holds the IP address
-	ip := request.PathParameter(ServiceNode)
+	key := request.Request.URL.String()
+	out, cached := ds.cache[key]
+	if ds.disableCache || !cached {
+		if sc := request.PathParameter(ServiceCluster); sc != ds.mesh.IstioServiceCluster {
+			errorResponse(response, http.StatusNotFound,
+				fmt.Sprintf("Unexpected %s %q", ServiceCluster, sc))
+			return
+		}
+		// service-node holds the IP address
+		ip := request.PathParameter(ServiceNode)
 
-	// route-config-name holds the listener port
-	routeConfigName := request.PathParameter(RouteConfigName)
-	port, err := strconv.Atoi(routeConfigName)
-	if err != nil {
-		errorResponse(fmt.Sprintf("Unexpected %s %q", RouteConfigName, routeConfigName), response)
-		return
-	}
+		// route-config-name holds the listener port
+		routeConfigName := request.PathParameter(RouteConfigName)
+		port, err := strconv.Atoi(routeConfigName)
+		if err != nil {
+			errorResponse(response, http.StatusNotFound,
+				fmt.Sprintf("Unexpected %s %q", RouteConfigName, routeConfigName))
+			return
+		}
 
-	instances := ds.services.HostInstances(map[string]bool{ip: true})
-	services := ds.services.Services()
-	httpRouteConfigs := buildOutboundHTTPRoutes(instances, services, &ProxyContext{
-		Discovery:  ds.services,
-		Config:     ds.config,
-		MeshConfig: ds.mesh,
-		IPAddress:  ip,
-	})
+		instances := ds.services.HostInstances(map[string]bool{ip: true})
+		services := ds.services.Services()
+		httpRouteConfigs := buildOutboundHTTPRoutes(instances, services, &ProxyContext{
+			Discovery:  ds.services,
+			Config:     ds.config,
+			MeshConfig: ds.mesh,
+			IPAddress:  ip,
+		})
 
-	routeConfig, ok := httpRouteConfigs[port]
-	if !ok {
-		errorResponse(fmt.Sprintf("Missing route config for port %d", port), response)
-		return
+		routeConfig, ok := httpRouteConfigs[port]
+		if !ok {
+			errorResponse(response, http.StatusNotFound,
+				fmt.Sprintf("Missing route config for port %d", port))
+			return
+		}
+		if out, err = json.MarshalIndent(routeConfig, " ", " "); err != nil {
+			errorResponse(response, http.StatusInternalServerError, err.Error())
+			return
+		}
+		ds.cache[key] = out
 	}
+	ds.cacheHitMiss(key, cached)
+	writeResponse(response, out)
+}
 
-	if err = response.WriteEntity(routeConfig); err != nil {
+func errorResponse(r *restful.Response, status int, msg string) {
+	glog.Warning(msg)
+	if err := r.WriteErrorString(status, msg); err != nil {
 		glog.Warning(err)
 	}
 }
 
-func errorResponse(msg string, response *restful.Response) {
-	glog.Warning(msg)
-	if err := response.WriteErrorString(404, msg); err != nil {
+func writeResponse(r *restful.Response, data []byte) {
+	r.WriteHeader(http.StatusOK)
+	r.Header().Set("Content-Type", "application/json")
```

Comment:
Let's use this https://github.com/emicklei/go-restful/blob/master/examples/restful-user-resource.go#L45 instead.